### ISSUE TITLE
[13.0] ir.model.data for mixin field not created

### ISSUE THREAD

**Issue by beledouxdenis** [odoo/odoo#49354]
### Impacted versions:
13.0

### Steps to reproduce:
Install a 13.0 from scratch with the `website_event` module

### Current behavior:
There is no `ir.model.data` record created for the field `is_published` of `event.event`.
```sql
13.0=# SELECT * FROM ir_model_data WHERE model = 'ir.model.fields' AND res_id = (select id from ir_model_fields where model = 'event.event' and name = 'is_published');
 id | create_uid | create_date | write_date | write_uid | noupdate | name | date_init | date_update | module | model | res_id 
----+------------+-------------+------------+-----------+----------+------+-----------+-------------+--------+-------+--------
(0 rows)
```
This has multiple incidences:
 1. When uninstalling the module `website_event`, the column `is_published` is not removed from the table `event_event` while it should.
 2. More problematic, when upgrading a 12.0 database to 13.0, the column ` is_published` is dropped from the table `event_event`, because of `_process_end` which deletes `ir.model.data` and their associated record if the xmlid wasn't loaded during the module update. As there is no xmlid for this field in 13.0, the ORM considers the `ir.model.data`, which is there in the 12.0 database, no longer exists in the source code, and decide to delete the `ir.model.data`, along with its related `ir.model.fields`, along with its table column, therefore deleting data during the upgrade :>.
https://github.com/odoo/odoo/blob/55ffc3027d26a36fc390d5efefa2d743cdc3e09f/odoo/addons/base/models/ir_model.py#L1941-L1948

### Expected behavior:
A `ir.model.data` record is created for the field ìs_published` of `event.event`. This is actually the case in 12.0.
```sql
12.0=# SELECT * FROM ir_model_data WHERE model = 'ir.model.fields' AND res_id = (select id from ir_model_fields where model = 'event.event' and name = 'is_published');
  id  | create_uid | create_date | write_date | write_uid | noupdate |              name               |         date_init          |        date_update         |    module     |      model      | res_id 
------+------------+-------------+------------+-----------+----------+---------------------------------+----------------------------+----------------------------+---------------+-----------------+--------
 6402 |            |             |            |           | f        | field_event_event__is_published | 2020-04-09 16:42:01.194264 | 2020-04-09 16:42:01.194264 | website_event | ir.model.fields |   2717
(1 row)
```

**Comment by beledouxdenis** [comment]
Not sure if this is for @nim-odoo or @rco-odoo or @Elkasitu , but this is relatively critical for upgrades, as users might loose data during upgrades without noticing it, so we would appreciate a prompt fix :D.

**Comment by nim-odoo** [comment]
Task 2235368 so it's in someone's pipe.

**Comment by nim-odoo** [comment]
Not sure how we should handle this. As @agr-odoo noticed, in v12 the field is defined on `event.event`:

https://github.com/odoo/odoo/blob/a699e408274a9d919214912faa1688b1d66fcefd/addons/website_event/models/event.py#L34

To compare both versions:
```
[v13]
field_website_published_mixin__is_published
field_website_published_multi_mixin__is_published
field_website_page__is_published
field_res_partner__is_published
field_res_users__is_published

[V12]
field_website_published_mixin__is_published
field_website_published_multi_mixin__is_published
field_website_page__is_published
field_event_event__is_published
```

@rco-odoo @Elkasitu What do you think?

**Comment by agr-odoo** [comment]
After further investigation probably the solution could be match the actual field 'website_published'
https://github.com/odoo/odoo/blob/5ff1e41c98f042fd13a1762977cc76604231dfd4/addons/website_event/models/event.py#L26
with a 'is_published' to make the model data match v12, by renaming or with a related field. I guess this could not be done in stable, so it should be added in the migration scripts

**Comment by beledouxdenis** [comment]
I see no reason why no ir.model.data would be created for a field.

That's how uninstallation work:
 - Each time a field is mentioned in a module, whether its a new field or an override, an ir.model.data for the field, with the according module, is created,
 - During the module uninstall, the ir.model.data of the field for that module is removed. If there is no other ir.model.data pointing to the field, then, and only then, the ir.model.fields record is removed, and the column in the table as well. if there is another module still using the field, there should still be another ir.model.data pointing to the field, therefore preventing the field to be removed.

With the current behavior, the field is just never removed when the module is uninstalled, which is not normal, in my opinion.

This is basically to handle the remove of fields for modules defining a same field on a same model, but without dependency between these modules.

**Comment by nim-odoo** [comment]
If we compare 12.0 and 13.0, and apply the following diff:

https://gist.github.com/a911d6d2454e3c1e638f1b5ea83798cf

we have the following entries matching the query:

```
select name from ir_model_data where name like '%is_published' order by name;
```

12.0
```
field_website_page__is_published
field_website_published_mixin__is_published
field_website_published_multi_mixin__is_published
```

13.0
```
field_res_partner__is_published
field_res_users__is_published
field_website_page__is_published
field_website_published_mixin__is_published
field_website_published_multi_mixin__is_published
```

The entry `field_event_event__is_published` is created in 12.0 because the field is defined explicitly on the model.

I don't know how upgrades and uninstallations were handled for such fields up to now, because `is_published` is not an exception. Indeed, other fields are impacted such as `website_meta_keywords`. In 12.0 and 13.0:
```
field_ir_ui_view__website_meta_keywords
field_website_page__website_meta_keywords
field_website_seo_metadata__website_meta_keywords
```

No entry related to `event_event` is created in `ir_model_data` while the field is physically on the `event_event` table.

Note that in this case, the mixin inherits from other mixins:
https://github.com/odoo/odoo/blob/b06fed46bb08a09ed56569e3f7c20e0908fc4c00/addons/website/models/mixins.py#L185-L186

Still @rco-odoo opinion would be helpful here on the direction a fix should take, and in which version it should be applied.

**Comment by Elkasitu** [comment]
Hi,

XMLIDs are only created for a field F of model M (whether a mixin or not) for the modules in which the field is defined / extended.

In other words, any other models inheriting from M within the same module (in this case, website) will create an xmlid for field F, but models from any other module will not generate xmlids for the same fields unless these fields are extended.

The difference between the two versions, as @agr-odoo noticed, is that in v12 the field is explicitly overridden inside `website.event`, this added the module `website_event` to the `_modules` attribute of field F, which in turn allows the creation of the `field_event_event__is_published` xmlid for the `website_event` module when it is installed / upgraded.

I talked to @rco-odoo about it and he doesn't seem to think it's a bug with the way xmlid are created and that it's rather more of a misconception of how mixins work or are supposed to work, he's ok with @nim-odoo's patch proposal

**Comment by beledouxdenis** [comment]
> he's ok with @nim-odoo's patch proposal
Which patch?

> The difference between the two versions, as @agr-odoo noticed, is that in v12 the field is explicitly overridden inside website.event,

I understood that. We can maybe do an exception for this specific field in the upgrade script.

Though, despite the upgrade issue, there is also the uninstall issue, in standard.

Is that a normal behavior the column is not removed from database when the module is uninstalled? I do not think so.

It would mean the data is kept when uninstalling and then reinstalling a module, which is known to be error-prone.

**Comment by Elkasitu** [comment]
> Is that a normal behavior the column is not removed from database when the module is uninstalled? I do not think so.

It is not, but in this case it's due to "api misuse" if you will, the fix in master would be to make website_event extend `is_published` so that the xmlid is created and the field / column deleted.

Although I suppose if it is an api misuse maybe we should enforce it or have better documentation, as it is it's a silent bug waiting to happen

I guess in stable an uninstall_hook that takes care of the `ir.model.field` would be enough

> Which patch?

I misread his gist, see above for fix

**Comment by agr-odoo** [comment]
https://github.com/odoo/odoo/pull/49533
(maybe some additional check is needed to avoid deleting data still referenced ?)

**Comment by beledouxdenis** [comment]
@Elkasitu 
I do not follow why you say this is a mis-use of the ORM.

I replicated the issue, using new modules with `odoo scaffold`, to simplify the use case:

`my_module/models/models.py`
```python
from odoo import models, fields, api


class my_module(models.Model):
    _name = 'my_module.my_module'
    _description = 'my_module.my_module'

    name = fields.Char()
    value = fields.Integer()
    value2 = fields.Float(compute="_value_pc", store=True)
    description = fields.Text()

    @api.depends('value')
    def _value_pc(self):
        for record in self:
            record.value2 = float(record.value) / 100
```
`website_my_module/models/models.py`
```python
from odoo import models


class my_module(models.Model):
    _name = 'my_module.my_module'
    _inherit = ['my_module.my_module', 'website.published.mixin']
```

With this, the `ir.model.data` for the `is_published` is not created:
```sql
13.0=# select * from ir_model_data where name ilike '%is_published%';
  id  | create_uid | create_date | write_date | write_uid | noupdate |                       name                        |         date_init          |        date_update         | module  |      model      | res_id 
------+------------+-------------+------------+-----------+----------+---------------------------------------------------+----------------------------+----------------------------+---------+-----------------+--------
 7313 |            |             |            |           | f        | field_website_published_mixin__is_published       | 2020-04-14 14:22:23.486746 | 2020-04-14 14:22:23.486746 | website | ir.model.fields |   2663
 7322 |            |             |            |           | f        | field_website_published_multi_mixin__is_published | 2020-04-14 14:22:23.486746 | 2020-04-14 14:22:23.486746 | website | ir.model.fields |   2671
 7404 |            |             |            |           | f        | field_website_page__is_published                  | 2020-04-14 14:22:23.486746 | 2020-04-14 14:22:23.486746 | website | ir.model.fields |   2747
 7490 |            |             |            |           | f        | field_res_partner__is_published                   | 2020-04-14 14:22:23.486746 | 2020-04-14 14:22:23.486746 | website | ir.model.fields |   2819
 7498 |            |             |            |           | f        | field_res_users__is_published                     | 2020-04-14 14:22:23.486746 | 2020-04-14 14:22:23.486746 | website | ir.model.fields |   2825
(5 rows)
```

While the column is well there in the table:
```sql
13.0=# \d my_module_my_module
                                         Table "public.my_module_my_module"
    Column    |            Type             | Collation | Nullable |                     Default                     
--------------+-----------------------------+-----------+----------+-------------------------------------------------
 id           | integer                     |           | not null | nextval('my_module_my_module_id_seq'::regclass)
 name         | character varying           |           |          | 
 value        | integer                     |           |          | 
 value2       | double precision            |           |          | 
 description  | text                        |           |          | 
 create_uid   | integer                     |           |          | 
 create_date  | timestamp without time zone |           |          | 
 write_uid    | integer                     |           |          | 
 write_date   | timestamp without time zone |           |          | 
 is_published | boolean                     |           |          |
```

Could you explain why you believe this is a mis-use of the ORM?

For an easier replication, here is the zip containing the modules I used in my above explanations:
[my_modules.zip](https://github.com/odoo/odoo/files/4475865/my_modules.zip)

**Comment by Elkasitu** [comment]
@beledouxdenis Let me preface this by saying I do not know the implementation details nor the limitations of mixins.

With that in mind, I talked to @rco-odoo yesterday morning, showed him the issue because to me it seemed weird too (despite knowing the implementation details of how xmlids are generated and ofc how uninstallation works), he said the bit about the XMLID not being created was normal so I *assumed* that either a) mixins are limited to the module in which they're defined or b) when they do, the inherited members must be overridden in order to generate an XMLID, but ofc I could be wrong.

If either of my assumptions were right, then it'd be a case of API misuse.

I'll ask him again and maybe to give an official statement to avoid any further misunderstandings :sweat_smile:

**Comment by beledouxdenis** [comment]
For information:
1. In 11.0, it was already working better, though it was not perfect. Using the same module than in https://github.com/odoo/odoo/issues/49354#issuecomment-613478043, then:
    ```sql
    11.0=# select * from ir_model_data where name ilike '%my_module%published%';
      id  | create_uid | create_date | write_date | write_uid | noupdate |                    name                     |         date_init          |        date_update         | module  |      model      | res_id 
    ------+------------+-------------+------------+-----------+----------+---------------------------------------------+----------------------------+----------------------------+---------+-----------------+--------
     5415 |            |             |            |           |          | field_my_module_my_module_website_published | 2020-04-15 08:21:15.807449 | 2020-04-15 08:21:15.807449 | website | ir.model.fields |   2206
    (1 row)
    ```
    While, from 12.0, with the same module:
    ```sql
    12.0=# select * from ir_model_data where name ilike '%my_module%published%';
     id | create_uid | create_date | write_date | write_uid | noupdate | name | date_init | date_update | module | model | res_id 
    ----+------------+-------------+------------+-----------+----------+------+-----------+-------------+--------+-------+--------
    (0 rows)
    ```
    So the field and its column were removed at uninstall of the `website`module (which is already better than now), and this is also probably the reason why we did not had this column being dropped  during upgrades up to know, because there was at least one `ir.model.data` pointing to the field, even though the module of that ` ir.model.data` was not accurate.

2. The problematic line, deciding if the `ir.model.data`should be created or not for a field being created, is this one:
https://github.com/odoo/odoo/blob/5c67b88b3655d61565853917e59cbf3e511852cd/odoo/addons/base/models/ir_model.py#L868

To me, this is a bug, occuring since Odoo 12.0. There is probably also a bug in 11.0, as it drops the column at the uninstall of `website` instead of  `website_event`,  but its less critical.

Maybe, one solution would consist to slightly alter the above condition to create the  `ir.model.data` if the field comes from an `AbstractModel` rather than a `BaseModel`.

**Comment by rco-odoo** [comment]
@beledouxdenis I understand the issue, but here are some elements to consider.

The XMLIDs of fields are created by method `_reflect_model` of `ir.model.fields`.  This method is **identical** in versions 12.0 and 13.0.  Therefore, this issue **is not new**, and clearly **not a regression**.  That being said, I am willing to find a solution for that nasty problem!

The criterion for adding an XMLID to a field in a given module is:
- the module is the one that first introduces the model (`model._original_module`), or
- the module is one of those that defined the field (`field._modules`).

The trick here is that for `event.event.is_published` in Odoo 13.0:
- the model is first introduced by module `event`,
- the field is defined by module `website` (and not overridden by any other module),
- the field is added on the model by inheritance in module `website_event`.

We can see that we would need an XMLID for the field in module `website_event`.  But currently I see no simple way of expressing "which module introduced field X by inheritance"...

You think this was enough?  Well, no!  The problem is **actually more complex than you imagine**.  Consider a slightly more complex scenario where:
- module A defines model M;
- module B defines mixin model X;
- module C depends on A and B, and makes M inherit from X;
- module B1 depends on B, and adds field F on X.

In this situation, uninstalling B1 clearly must remove F.  But uninstalling C should also remove F.  In this case, the presence of field F depends on **both** modules B1 and C to be installed.  We simply cannot implement the removal of field F based on its XMLIDs in this situation.  The current mechanism deals with a situation were F depends on B1 **or** C to be installed, but not on B1 **and** C to be installed.  We actually need another mechanism in this case!

@Elkasitu @nim-odoo @agr-odoo

**Comment by beledouxdenis** [comment]
How about my point regarding 11.0? That it was already better as situation?
https://github.com/odoo/odoo/issues/49354#issuecomment-613955231
> Therefore, this issue is not new, and clearly not a regression. 
Well, compared to 11.0, it is.

I can understand that for the case of the `is_published` in `website_event`, it's a complex case we can maybe ignore, but what about the simple case, illustrated with the modules I attached in a previous comment?
The behavior is not the same between 11.0 and 12.0/13.0.

What do you suggest?
 - For the upgrade:
   It's like impossible to detect this by just reading the diff code between two versions. Therefore this will happen again. In this case, the information loss is not critical, but it could be for the next case.
We can make a fix post-morterm, including for this `is_published` case, but what about the future?

 - For the uninstall:
    An uninstall hook for each case?

I do not say it has to be solved in 13.0. We can make a deal to solve this manually in upgrade script from now, there is probably no other cases. But a solution must be found for master, at least.

**Comment by rco-odoo** [comment]
> > Therefore, this issue is not new, and clearly not a regression.
> 
> Well, compared to 11.0, it is.

I am sorry, you are right.  I thought the problem was 12.0 vs 13.0, and did not check in 11.0.

The implementation of Odoo 11.0 had a much simpler criterion for when to add an XMLID: *at creation*, period.  And this was buggy in some cases (two modules introducing the same field).  That's why we came with these new criteria, that we thought were correct.  You indeed proved us wrong.

> I can understand that for the case of the `is_published` in `website_event`, it's a complex case we can maybe ignore, but what about the simple case, illustrated with the modules I attached in a previous comment?
> The behavior is not the same between 11.0 and 12.0/13.0.

To be more precise: the behavior is the same _if you don't introduce new parent models in model extensions_, which represent _most cases_.  When you do, the behavior is different indeed.

> What do you suggest?
> 
> * For the upgrade:
>   It's like impossible to detect this by just reading the diff code between two versions. Therefore this will happen again. In this case, the information loss is not critical, but it could be for the next case.
>   We can make a fix post-morterm, including for this `is_published` case, but what about the future?
> * For the uninstall:
>   An uninstall hook for each case?
> 
> I do not say it has to be solved in 13.0. We can make a deal to solve this manually in upgrade script from now, there is probably no other cases. But a solution must be found for master, at least.

There is no point in fixing this in 13.0, as we will have to deal with existing databases, where fields already don't have an XMLID.

We will try to solve this in master, I just created a task for it.

**Comment by Yenthe666** [comment]
I believe this can be closed as it was handled in https://github.com/odoo/odoo/commit/535ae09b48b285e0d881bd2f81021b21c660cb5e for 13.4
